x-pack/plugins/observability_solution/investigate_app/public/pages/details/contexts/investigation_context.tsx

```diff
@@ -0,0 +1,230 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { i18n } from '@kbn/i18n';
+import { type GlobalWidgetParameters } from '@kbn/investigate-plugin/public';
+import { GetInvestigationResponse, InvestigationItem, Item } from '@kbn/investigation-shared';
+import { isEqual } from 'lodash';
+import React, { createContext, useContext, useEffect, useRef, useState } from 'react';
+import { useAddInvestigationItem } from '../../../hooks/use_add_investigation_item';
+import { useAddInvestigationNote } from '../../../hooks/use_add_investigation_note';
+import { useDeleteInvestigationItem } from '../../../hooks/use_delete_investigation_item';
+import { useDeleteInvestigationNote } from '../../../hooks/use_delete_investigation_note';
+import { useFetchInvestigation } from '../../../hooks/use_fetch_investigation';
+import { useKibana } from '../../../hooks/use_kibana';
+import { useUpdateInvestigation } from '../../../hooks/use_update_investigation';
+import { useUpdateInvestigationNote } from '../../../hooks/use_update_investigation_note';
+
+export type RenderedInvestigationItem = InvestigationItem & {
+  loading: boolean;
+  element: React.ReactNode;
+};
+
+interface InvestigationContextProps {
+  investigation?: GetInvestigationResponse;
+  renderableItems: RenderedInvestigationItem[];
+  globalParams: GlobalWidgetParameters;
+  updateInvestigationParams: (params: GlobalWidgetParameters) => Promise<void>;
+  addItem: (item: Item) => Promise<void>;
+  deleteItem: (itemId: string) => Promise<void>;
+  isAddingItem: boolean;
+  isDeletingItem: boolean;
+  // note
+  addNote: (content: string) => Promise<void>;
+  updateNote: (noteId: string, content: string) => Promise<void>;
+  deleteNote: (noteId: string) => Promise<void>;
+  isAddingNote: boolean;
+  isUpdatingNote: boolean;
+  isDeletingNote: boolean;
+}
+
+export const InvestigationContext = createContext<InvestigationContextProps>({
+  renderableItems: [],
+  globalParams: { timeRange: { from: '', to: '' } },
+  investigation: undefined,
+  updateInvestigationParams: async () => {},
+  addItem: async () => {},
+  deleteItem: async () => {},
+  isAddingItem: false,
+  isDeletingItem: false,
+  // note
+  addNote: async () => {},
+  updateNote: async (noteId: string, content: string) => {},
+  deleteNote: async (noteId: string) => {},
+  isAddingNote: false,
+  isUpdatingNote: false,
+  isDeletingNote: false,
+});
+
+export function useInvestigation() {
+  return useContext(InvestigationContext);
+}
+
+export function InvestigationProvider({
+  initialInvestigation,
+  children,
+}: {
+  initialInvestigation: GetInvestigationResponse;
+  children: React.ReactNode;
+}) {
+  const {
+    dependencies: {
+      start: { investigate },
+    },
+  } = useKibana();
+
+  const { data: investigation, refetch } = useFetchInvestigation({
+    id: initialInvestigation.id,
+    initialInvestigation,
+  });
+
+  const cache = useRef<
```

Comment:
Nice implementing caching here to avoid unnecessary re-renders. 

I was wondering if we should move the `rendering`logic outside of the `Context`. Like creating an `<ItemRenderer>` component

Reply:
I think it's easier if it's collocated to the other context state. But we could definitely extract the logic to a hook and use it form this context.